### PR TITLE
Fix (undocumented) issue with hat presses not registering.

### PR DESCRIPTION
**Pull request by z64555**
Need some run-tests to verify operation.

### CODE REVIEW

**Comment by z64555** [comment]
The issue arose from changes made in PR #671

File: code/controlconfig/controlsconfigcommon.cpp [hunk 1]
```diff
@@ -612,6 +613,9 @@ void control_config_common_init()
 		Scan_code_text = Scan_code_text_english;
 		Joy_button_text = Joy_button_text_english;
 	}
+
+	auto current = io::joystick::getCurrentJoystick();
```
Comment (asarium) [inline code/controlconfig/controlsconfigcommon.cpp]: `getCurrentJoystick` can return `nullptr` if there is no joystick. #773 fixed a similar issue so you should make sure that `current` is not `nullptr` when you use it.

File: code/io/joy-sdl.cpp [hunk 2]
```diff
@@ -364,28 +393,24 @@ namespace joystick
 		return _id;
 	}
 
-	void Joystick::fillValues()
-	{
+	void Joystick::fillValues() {
```
Comment (asarium) [inline code/io/joy-sdl.cpp]: There are a bunch of format changes here that result in no functional difference.

File: code/io/joy.h [hunk 3]
```diff
@@ -35,6 +35,8 @@ namespace io
 			HAT_RIGHTDOWN,
 			HAT_LEFTUP,
 			HAT_LEFTDOWN,
+
+			HAT_NUM_POS     // This one is always last
```
Comment (asarium) [inline code/io/joy.h]: If you are using the enum length trick then the first element must have the value 0 or the indices may be invalid.

File: code/io/joy.h [hunk 4]
```diff
@@ -214,8 +238,11 @@ namespace io
 			SCP_vector<coord2d> _ballValues; //!< The ball values
 			SCP_vector<HatPosition> _hatValues; //!< The current hat values
 
-			SCP_vector<int> _buttonDownTimestamp; //!< The timestamp since when the button is pressed, -1 if not pressed
-			SCP_vector<int> _buttonDownCount; //!< The number of times the button was pressed
+			SCP_vector<int> _buttonDownTimestamp;       //!< The timestamp since when the button is pressed, -1 if not pressed
+			SCP_vector<int> _buttonDownCount;           //!< The number of times the button was pressed
+			SCP_vector<int> _hatDownTimestamp;          //!< The timestamp when the hat activated, -1 if inactive (centered)
+			SCP_vector<SCP_vector<int>> _hatDownCount;  //!< The number of times each hat position has been hit [hat][hatpos]
```
Comment (asarium) [inline code/io/joy.h]: Could you put the various hat values in its own struct? That would make it clearer which values belong together.

**Comment by z64555** [comment]
Adressed the issues pointed out by m|m, and put the button info and hat info into their respective array of structures

**Comment by z64555** [comment]
Welp. need to fix the config labels

**Comment by z64555** [comment]
So, in case the Team Fortress Heavy accented message is indecipherable, we can't have dynamic number of buttons about 32 yet. Nor can we have more than one hat for the time being.

These two are features and is outside the scope of this bugfix PR.

...and would require more work.

**Comment by asarium** [comment]
The code looks good. However, I can't assign a binding to the hat on my joystick. I'm going to try to find out why this happens.

File: code/io/joy-sdl.cpp [hunk 5]
```diff
@@ -9,15 +9,66 @@
...
 Joystick *currentJoystick = nullptr;
 
 bool initialized = false;
 
+/**
+ * @brief Compatibility conversion from HatPosition to array index
+ */
+inline
+int hatEnumToIdx(HatPosition in) {
+	return int(in - JOY_NUM_BUTTONS);
```
Comment (asarium) [inline code/io/joy-sdl.cpp]: Minor nitpick: `static_cast` or a C-style cast should be used here. Function style casts are potentially dangerous. This one isn't but it's better to keep the code consistent.

File: code/io/joy-sdl.cpp [hunk 6]
```diff
@@ -9,15 +9,66 @@
...
+
+/**
+* @brief Compatibility conversion from array index to HatPosition
+*/
+inline
+HatPosition hatIdxToEnum(int in) {
+	switch (in) {
+	case 0:
+		return HAT_DOWN;
+		break;
```
Comment (asarium) [inline code/io/joy-sdl.cpp]: The `break` is not needed.

File: code/io/joy-sdl.cpp [hunk 7]
```diff
@@ -351,7 +402,11 @@ namespace joystick
 			// Hat is inactive
 			return 0.0f;
 
-		} // Hat is active
+		} if (_hat[index].Value != hatIdxToEnum(pos)) {
```
Comment (asarium) [inline code/io/joy-sdl.cpp]: Should this be an `else if`? If it shouldn't then the `if` should not be on the same line as the closing brace.

**Comment by z64555** [comment]
Ok, amended.

**Pull request merged**